configure.ac

```diff
@@ -313,22 +313,100 @@ if test "x$enable_werror" = "xyes"; then
-  AX_CHECK_COMPILE_FLAG([-Wself-assign],[CXXFLAGS="$CXXFLAGS -Wno-self-assign"],,[[$CXXFLAG_WERROR]])
-  AX_CHECK_COMPILE_FLAG([-Wunused-local-typedef],[CXXFLAGS="$CXXFLAGS -Wno-unused-local-typedef"],,[[$CXXFLAG_WERROR]])
-  AX_CHECK_COMPILE_FLAG([-Wdeprecated-register],[CXXFLAGS="$CXXFLAGS -Wno-deprecated-register"],,[[$CXXFLAG_WERROR]])
-fi
+
+# This enables all warnings and selectively disables some. Individual warnings should be
+# 1. fixed,
+# 2. locally silenced with an explanatory comment, or
+# 3. globally silenced here with an explanatory comment.
+#
+# Justifications
+# • `-Wno-aix-compat`: we don’t want to maintain compatibility with IBM AIX compilers
+# • `-Wno-c++-compat`: this warns about an ABI layout incompatibility between C and C++,
+#    but we don’t need such compatibility except where we use `extern "C"`
+# • `-W*c++*-compat-pedantic`: disables warnings about compatibility with older versions of C++ and
+#    enables warnings about compatibility with newer versions of C++ (we need the explicit enabling,
+#    even with `-Weverything`, because the earlier disabling of older versions removes some warnings
+#    that we do want)
+DESIRED_WARNING_CXXFLAGS="\
+  --system-header-prefix=boost --system-header-prefix=crypto/ctaes --system-header-prefix=db_cxx \
+  --system-header-prefix=event2 --system-header-prefix=gmock --system-header-prefix=gtest \
+  --system-header-prefix=leveldb --system-header-prefix=rust --system-header-prefix=sodium \
+  --system-header-prefix=univalue --system-header-prefix=zmq \
+  -Weverything \
+  -Wno-aix-compat -Wno-c++-compat \
+  -Wno-c++98-compat-pedantic -Wno-c++11-compat-pedantic -Wno-c++14-compat-pedantic \
+  -Wc++17-compat-pedantic -Wc++20-compat-pedantic"
+
+# This list should be incrementally pruned. Each flag should either be enabled or moved to
+# `DESIRED_WARNING_CXXFLAGS` with an explanation as to why it remains disabled.
+#
+# Guidelines for fixing warnings:
+# • fix a small number (perhaps one) at a time
+# • remove `-Werror` locally to get the full report of the warning you enabled (but be careful, as
+#   re-running `make` will not recompile files that only warned).
+#
+# Potential justifications:
+# • `-Wno-format-nonliteral`: Bitcoin i18n
+# • `-Wno-old-style-cast`: do after `-W*-conversion` to avoid introducing more conversions
+# • `-Wno-padded`: is this worth the noise of rearranging fields & adding anonymous bitfields?
+# • `-Wno-shadow-field`: conflicts with our style of naming constructor arguments the same as fields
+# • `-Wno-shadow-field-in-constructor`: conflicts with our style of naming constructor arguments the
+#    same as fields
+# • `-Wno-string-conversion`: currently abused to `assert` on strings
+# • `-Wno-tautological-type-limit-compare`: I think isn’t portable because of numeric type sizing
+DISABLED_WARNING_CXXFLAGS="\
+  -Wno-c++20-compat -Wno-cast-align -Wno-cast-qual -Wno-comma -Wno-conditional-uninitialized \
```

Comment:
We really should make an effort to fix the warnings in [`-Wc++20-compat`](https://releases.llvm.org/15.0.0/tools/clang/docs/DiagnosticsReference.html#wc-20-compat), because those are blockers for moving to C++20.

Reply:
Instead of `-Wno-foo`, what about using `-Wno-error=foo`? That way you still get the warning but it doesn't break the build.

There might be certain warnings that are very noisy, or we don't agree they indicate a warnable issue, that we want to use `-Wno-foo` for. However, I think `-Wno-error=foo` should be the default.

Reply:
Hunh, odd … I saw your first comment here, but not the second in the thread (so I resolved this convo without having read it all).

I think the short answer to `-Wno-error=foo` is “maybe later”. With this PR, that would result in 10s of thousands of warnings (in part because Clang repeats all the warnings from headers every time they get included[^1]). The next PR cleans up most of the flags, but I’m not sure what percentage of the messages that eliminates. However, the flags that are left behind are the more troublesome ones and reporting warnings that _shouldn’t_ be fixed alongside other changes isn’t particularly helpful (especially when you have actual warnings/errors buried amongst the pre-existing noise).

There are a couple of other options – the best is just to get all the warnings turned into errors 😁 but barring that happening quickly, we could
1. silence the remaining warnings on a per-file basis, rather than globally – this prevents contributors from introducing warnings in files that didn’t have that particular warning before, and we can do finer-grained incremental fixing of warnings file-by-file instead of flag-by-flag or
2. have a separate linter that runs the build with `-Wno-error` and `DISABLED_WARNING_CXXFLAGS=""` – this will be a noisy linter that fails for a while, but longer-term I think this is a good approach for warnings that we consider “style warnings”.

Approach 1 is something I’ve done before along with a shrinking whitelist of files that were allowed to have warnings disabled (but I don’t think we need that part – our reviews should catch anyone adding `_Pragma("GCC diagnostic ignored \"-Wfoo\"")`).

Approach 2 is nice because inline warnings can often be missed when the build is successful overall. Seeing a failed check on a PR (even if not a required check) is more likely to get noticed.

[^1]: I wrote a little command-line tool, `dedup`, that replaces blocks of duplicated lines with a reference back to their first occurrence. `make | dedup` is quite helpful for this kind of thing.